### DISCUSSION TITLE
Python import fails when using tch-rs

### DISCUSSION THREAD

**Discussion by Kaiyotech** [PyO3/pyo3#3420] [Questions]
### Bug Description

If tch-rs is used in a rust project, Python cannot import that module. Rust side works fine and compiles and the test of the function works.

I've built the project by using Libtorch manual install, debug version (controlled by environment variable set to my debug path for LIBTORCH and for appending to PATH). I'm using the msvc toolchain. Attempting to build with the pytorch from the venv fails on the rust side, the Libtorch install seems more stable. My rust test is able to use the lib.rs successfully to print the Cuda device.

If I change the rust lib.rs file to not have `use tch;` and to print something else instead of torch it works fine in python and prints what I want to print.

### Steps to Reproduce

1. Create a simple example lib.rs in a pyo3 after maturin init on the python project (in my case called trch_test_py)
```
use pyo3::prelude::*;
use tch;

#[pyfunction]
pub fn test_cuda() {
    println!("{:?}", tch::Device::cuda_if_available());
}

#[pymodule]
fn trch_test_py(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(test_cuda, m)?)?;
    Ok(())
}
```

2. Maturin develop and then attempt to use this module from Python
```
import trch_test_py

trch_test_py.test_cuda()
```

### Backtrace

```
shell
Traceback (most recent call last):
  File "C:\Users\User\Code\Kaiyotech\trch_test_py\test.py", line 1, in <module>
    import trch_test_py
  File "C:\Users\User\Code\Kaiyotech\trch_test_py\venv\lib\site-packages\trch_test_py\__init__.py", line 1, in <module>
    from .trch_test_py import *
ImportError: DLL load failed while importing trch_test_py: The specified module could not be found.
```

### Your operating system and version

Windows 10 Home - 10.0.19045 N/A Build 19045

### Your Python version (`python --version`)

3.9.10

### Your Rust version (`rustc --version`)

rustc 1.72.0 (5680fa18f 2023-08-23)

### Your PyO3 version

0.19.2

### How did you install python? Did you use a virtualenv?

python.org installer I think, I've had this version for a long time so I'm not sure. I am using a venv.

**Comment by Kaiyotech** [comment]
Update: I'm able to build and use from Python https://github.com/LaurentMazare/tch-ext, which uses pyO3 0.18.3. I've tried to make my repo identical to his (including down grading pyo3 I think) but mine still fails and I'm not sure why. I'm guessing this is user error or perhaps a quirk in trch-rs? Let me know if you'd like me to close it and move it over there for troubleshooting. I wasn't sure if this was more pyo3 related or tch related.

**Reply by davidhewitt** [reply]
`C:\Users\User\Code\Kaiyotech\trch_test_py\venv\lib\site-packages\trch_test_py\__init__.py`

This looks like the wrong path to be importing Python source for your own repo, it's possible this got copied into the virtualenv alongside an old version of your project (or similar)? Does deleting and recreating the virtualenv resolve the issue?

I transferred this to a discussion because at the moment this looks to me like user error, though we might yet find a bug :)

**Comment by KarelPeeters** [comment]
I ran into a similar issue. In my case I'm using cuda directly instead of `tch`, but the error is exactly the same.

Using [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) I figured out which DLL files `python.exe` was trying to load, and I manually copied them over into the maturin `.env\Lib\site-packages\MY_LIBRARY` folder, next to the `.pyd` file. For me this was files like `cudnn64_8.dll`, `nvcuda.dll`, ... For you it's probably `libtorch.dll`, maybe also the cuda libraries.

This got everything working, at least when using `maturin develop`.

I don't yet know:
* What is the right way to do this? I can't find Anything in the [Maturin User Guide](https://www.maturin.rs/) on how to mark certain DLLs as required.
* How should the DLLs be packaged for distribution and eventual publishing to PyPI?
* Why does running bare Rust executables work, while loading the library in Python doesn't? Don't they both have the same DLL search paths (ie. current folder, `$PATH`, `C:\Windos\System32`, ...)?
